src/upchunk.ts

```diff
@@ -305,37 +346,55 @@ export class UpChunk  {
     this.getChunk()
       .then(() => {
         this.attemptCount = this.attemptCount + 1;
-
+        this.lastChunkStart = new Date();
         return this.sendChunk()
       })
       .then((res) => {
         if (SUCCESSFUL_CHUNK_UPLOAD_CODES.includes(res.statusCode)) {
+          const lastChunkEnd = new Date();
+          const lastChunkInterval = (lastChunkEnd.getTime() - this.lastChunkStart.getTime()) / 1000;
+
           this.dispatch('chunkSuccess', {
             chunk: this.chunkCount,
+            chunkSize: this.chunkSize,
             attempts: this.attemptCount,
+            timeInterval: lastChunkInterval,
             response: res,
           });
 
           this.attemptCount = 0;
           this.chunkCount = this.chunkCount + 1;
+          this.nextChunkRangeStart = this.nextChunkRangeStart + this.chunkByteSize;
 
           if (this.chunkCount < this.totalChunks) {
+            // dynamic chunk sizing
+            if (this.dynamicChunkSize) {
+              if (lastChunkInterval < 10)
+              {
+                this.chunkSize = Math.min(this.chunkSize * 2,this.maxChunkSize);
+              } else if (lastChunkInterval > 30) {
+                this.chunkSize = Math.max(this.chunkSize / 2,this.minChunkSize);
+              }
+
+              // Now update the new chunkByteSize to the newly calculated chunk size
+              this.chunkByteSize = this.chunkSize * 1024;
+
+              // Re-estimate the total number of chunks, by adding the completed
+              // chunks to the remaining chunks
+              const remainingChunks = (this.file.size-this.nextChunkRangeStart)/this.chunkByteSize;
+              this.totalChunks = Math.ceil(this.chunkCount + remainingChunks);
+            }
             this.sendChunks();
           } else {
             this.success = true;
             this.dispatch('success');
           }
 
-          const chunkFraction = this.chunkCount / this.totalChunks;
-          const uploadedBytes = chunkFraction * this.file.size;
-
-          const percentProgress = (100 * uploadedBytes) / this.file.size;
-
-          this.dispatch('progress', percentProgress);
         } else if (TEMPORARY_ERROR_CODES.includes(res.statusCode)) {
```

Comment:
On failures, should we reset the chunk size to the initial value?

Reply:
I'm not sure there's a reason to right now -- in my testing, a chunk would occasionally fail (maybe once out of every 200 chunks), and I never saw a chunk fail more than once.